clusterloader2/pkg/measurement/common/pod_startup_latency.go

```diff
@@ -0,0 +1,296 @@
+}
+
+func createPodStartupLatencyMeasurement() measurement.Measurement {
+	return &podStartupLatencyMeasurement{
+		createTimes:   make(map[string]metav1.Time),
+		scheduleTimes: make(map[string]metav1.Time),
+		runTimes:      make(map[string]metav1.Time),
+		watchTimes:    make(map[string]metav1.Time),
+	}
+}
+
+type podStartupLatencyMeasurement struct {
+	namespace     string
+	labelSelector string
+	fieldSelector string
+	informer      cache.SharedInformer
+	stopCh        chan struct{}
+	createTimes   map[string]metav1.Time
+	scheduleTimes map[string]metav1.Time
+	runTimes      map[string]metav1.Time
+	watchTimes    map[string]metav1.Time
+}
+
+// Execute supports two actions:
+// - start - Starts to observe pods and pods events.
+// - gather - Gathers and prints current pod latency data.
+func (p *podStartupLatencyMeasurement) Execute(config *measurement.MeasurementConfig) ([]measurement.Summary, error) {
+	var summaries []measurement.Summary
+	action, err := util.GetString(config.Params, "action")
+	if err != nil {
+		return summaries, err
+	}
+
+	switch action {
+	case "start":
+		p.namespace, err = util.GetStringOrDefault(config.Params, "namespace", metav1.NamespaceAll)
+		if err != nil {
+			return summaries, err
+		}
+		p.labelSelector, err = util.GetStringOrDefault(config.Params, "labelSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		p.fieldSelector, err = util.GetStringOrDefault(config.Params, "fieldSelector", "")
+		if err != nil {
+			return summaries, err
+		}
+		return summaries, p.start(config.ClientSet)
+	case "gather":
+		return p.gather(config.ClientSet)
+	default:
+		return summaries, fmt.Errorf("unknown action %v", action)
+	}
+
+}
+
+func (p *podStartupLatencyMeasurement) start(c clientset.Interface) error {
+	glog.Infof("Starting pod startup latency measurement...")
+	p.stopCh = make(chan struct{})
+	checkPod := p.checkPodFunc()
+	optionsModifier := func(options *metav1.ListOptions) {
+		options.FieldSelector = p.fieldSelector
+		options.LabelSelector = p.labelSelector
+	}
+	listerWatcher := cache.NewFilteredListWatchFromClient(c.CoreV1().RESTClient(), "pods", p.namespace, optionsModifier)
+	p.informer = cache.NewSharedInformer(listerWatcher, &corev1.Pod{}, 0)
+	p.informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			checkPod(obj)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			checkPod(newObj)
+		},
+	})
+
+	go p.informer.Run(p.stopCh)
+	timeoutCh := make(chan struct{})
+	timeoutTimer := time.AfterFunc(informerSyncTimeout, func() {
+		close(timeoutCh)
+	})
+	defer timeoutTimer.Stop()
+	if !cache.WaitForCacheSync(timeoutCh, p.informer.HasSynced) {
+		return fmt.Errorf("timed out waiting for caches to sync")
+	}
+	return nil
+}
+
+func (p *podStartupLatencyMeasurement) gather(c clientset.Interface) ([]measurement.Summary, error) {
+	glog.Infof("Gathering pod startup latency measurement...")
+	if p.informer == nil || p.stopCh == nil {
```

Comment:
This assumes that there is no concurrency.

Is that true, that we don't support that (if so, we should be pretty clear in some comemnts/docs).
If not, we should probably add a lock.

Reply:
I pretty sure that we don't want to have concurrency for the measurements. I will mention it in README.

Reply:
This should be mentioned also in this file (next to struct definition or its initialization or sth like that)